File: net/JetBrains.Serialization/src/FileInfos/Msi/MsiFileMetaInfo.cs
using System.Collections;
using JetBrains.FormatRipper.Compound;
using Newtonsoft.Json;

namespace JetBrains.Serialization.FileInfos.Msi;

[JsonObject(MemberSerialization.OptIn)]
public class MsiFileMetaInfo : IFileMetaInfo
{
  [JsonProperty("FileSize")] public long FileSize { get; set; }

  [JsonProperty("CompoundFileHeaderMetaInfo")]
  private CompoundFileHeaderMetaInfo CompoundFileHeaderMetaInfo { get; set; }

  [JsonProperty("Entries")] private List<CompoundFile.DirectoryEntry> Entries { get; set; }
  [JsonProperty("SpecialEntries")] private List<KeyValuePair<string, byte[]>> SpecialEntries { get; set; }
  [JsonProperty("SpecialSegments")] private List<KeyValuePair<long, byte[]>> SpecialSegments { get; set; }

  [JsonProperty("DigitalSignatureExData")]
  private byte[] DigitalSignatureExData { get; set; } // Nullable

  [JsonProperty("MiniStreamStartSector")]
  private int MiniStreamStartSector { get; set; }

  [JsonConstructor]
  public MsiFileMetaInfo(long fileSize, CompoundFileHeaderMetaInfo compoundFileHeaderMetaInfo,
    List<CompoundFile.DirectoryEntry> entries, List<KeyValuePair<string, byte[]>> specialEntries,
    List<KeyValuePair<long, byte[]>> specialSegments, byte[] digitalSignatureExData, int miniStreamStartSector)
  {
    FileSize = fileSize;
    CompoundFileHeaderMetaInfo = compoundFileHeaderMetaInfo;
    Entries = entries;
    SpecialEntries = specialEntries;
    SpecialSegments = specialSegments;
    DigitalSignatureExData = digitalSignatureExData;
    MiniStreamStartSector = miniStreamStartSector;
  }

  public void ModifyFile(Stream stream, byte[] signature)
  {
    var unsignedFile = CompoundFile.Parse(stream, CompoundFile.Mode.SIGNATURE_DATA);
    var unsignedEntries = unsignedFile.GetEntries();
    var unsignedEntriesMap = new Hashtable();
    foreach (var keyValuePair in unsignedEntries)
    {
      if (unsignedEntriesMap.Contains(keyValuePair.Key.Name.Trim(new[] { '' })))
        continue;
      unsignedEntriesMap.Add(keyValuePair.Key.Name.Trim(new[] { '' }), keyValuePair.Value);
    }

    var unsignedRoot = unsignedEntries.Find(
      entry => entry.Key.Name.Trim(new[] { '' }).Equals("Root Entry")
    );

    var startSect = unsignedRoot.Key.StartingSectorLocation;

    unsignedFile.PutEntries(unsignedEntries, (uint)startSect, wipe: true);

    unsignedFile = new CompoundFile(CompoundFileHeaderMetaInfo, stream);

    var specialEntriesDataMap = new Hashtable
    {
      { "MsiDigitalSignatureEx", DigitalSignatureExData },
      { "DigitalSignature", signature }
    };
    foreach (var kv in SpecialEntries)
    {
      specialEntriesDataMap.Add(kv.Key, kv.Value);
    }

    unsignedFile.PutEntries(
      Entries.Select(entry =>
        {
          var trimmedName = entry.Name.Trim(new[] { '' });
          if (specialEntriesDataMap.Contains(trimmedName))
            return new KeyValuePair<CompoundFile.DirectoryEntry, byte[]?>(entry,
              (byte[])specialEntriesDataMap[trimmedName]!);
          else
            return new KeyValuePair<CompoundFile.DirectoryEntry, byte[]?>(entry,
              (byte[])unsignedEntriesMap[trimmedName]!);
        }
      ).ToList()!, (uint)MiniStreamStartSector
    );
    //
    foreach (var segment in SpecialSegments)
    {
      stream.Position = segment.Key;
      stream.Write(segment.Value, 0, segment.Value.Length);
    }

    //
    if (FileSize < stream.Length)
    {
      stream.SetLength(FileSize);
    }
    else if (FileSize > stream.Length)
    {
      var diff = FileSize - stream.Length;
      stream.Write(new byte[diff], 0, (int)diff);
    }
  }
}

File: jvm/src/main/kotlin/com/jetbrains/signatureverifier/serialization/fileInfos/Msi/MsiFileMetaInfo.kt
package com.jetbrains.signatureverifier.serialization.fileInfos

import com.jetbrains.signatureverifier.cf.CompoundFile
import com.jetbrains.signatureverifier.cf.DirectoryEntry
import com.jetbrains.signatureverifier.cf.MsiFile
import com.jetbrains.signatureverifier.serialization.ByteArraySerializer
import com.jetbrains.signatureverifier.serialization.toHexString
import com.jetbrains.util.Jump
import kotlinx.serialization.Serializable
import java.nio.ByteBuffer
import java.nio.channels.SeekableByteChannel

@Serializable
class MsiFileMetaInfo(
  val fileSize: Long,
  private val compoundFileMetaInfo: CompoundFile.Companion.CompoundFileMetaInfo,
  private val entries: List<DirectoryEntry>,
  private val specialEntries: List<Pair<String, @Serializable(ByteArraySerializer::class) ByteArray>>,
  private val specialSegments: List<Pair<Int, @Serializable(ByteArraySerializer::class) ByteArray>>,
  @Serializable(ByteArraySerializer::class)
  private val digitalSignatureExData: ByteArray?,
  private val miniStreamStartSector: Int
) : FileMetaInfo {
  companion object {
    // Those entries vary between signed and unsigned files, but are not logically related to signature
    val knownSpecialEntryNames = listOf("40483F3BF2433844B145", "40483F3F77456C446A3EB2442F48")
  }

  override fun modifyFile(stream: SeekableByteChannel, signature: ByteArray) {
    var unsignedFile = MsiFile(stream)

    val unsignedEntries = unsignedFile.getEntries()
    val unsignedEntriesMap =
      unsignedEntries
        .associateBy { it.first.Name.toHexString() }

    val startSect = unsignedEntriesMap[MsiFile.rootEntryName.toHexString()]!!.first.StartSect

    unsignedFile.putEntries(
      unsignedEntries,
      startSect.toInt(),
      wipe = true
    )

    unsignedFile = MsiFile(compoundFileMetaInfo, stream)

    val specialEntriesDataMap = mutableMapOf(
      MsiFile.msiDigitalSignatureExEntryName.toHexString() to digitalSignatureExData,
      MsiFile.digitalSignatureEntryName.toHexString() to signature
    )
    specialEntriesDataMap.putAll(specialEntries)


    unsignedFile.putEntries(
      entries.map { entry ->
        entry.Name.toHexString().let { name ->
          specialEntriesDataMap[name]?.let { Pair(entry, it) } ?: Pair(entry, unsignedEntriesMap[name]!!.second)
        }
      },
      miniStreamStartSector
    )

    specialSegments.forEach {
      stream.Jump(it.first.toUInt())
      stream.write(ByteBuffer.wrap(it.second))
    }

    if (fileSize < stream.size()) {
      stream.truncate(fileSize)
    } else if (fileSize > stream.size()) {
      stream.write(ByteBuffer.wrap(ByteArray((fileSize - stream.size()).toInt())))
    }
  }
}